art/attacks/HCLU.py

```diff
@@ -0,0 +1,118 @@
+
+    def __init__(self, classifier, conf=0.95, unc_increase=100.0, minVal=0.0, maxVal=1.0):
+        """
+        :param classifier: A trained model of type scikit decision tree.
+        :type classifier: :class:`.Classifier.GPyGaussianProcessClassifier
+        :param conf: Confidence that examples should have, if there were to be classified as 1.0 maximally
+        :type conf: :float: 
+        :param unc_increase: Value uncertainty is allowed to deviate, where 1.0 is original value
+        :type unc_increase: :float:
+        :param minVal: minimal value any feature can take
+        :type minVal: :float:
+        :param maxVal: maximal value any feature can take
+        :type maxVal: :float:
+        """
+        self.classifier = classifier
+        if not isinstance(classifier, GPyGaussianProcessClassifier):
+            raise TypeError('Model must be a GPy Gaussian Process classifier!')
+        self.conf = conf
+        self.unc_increase = unc_increase
+        self.minVal = minVal
+        self.maxVal = maxVal
+
+    def generate(self, x, y=None):
+        """
+        Generate adversarial examples and return them as an array. This method should be overridden by all concrete
+        attack implementations.
+
+        :param x: An array with the original inputs to be attacked.
+        :type x: `np.ndarray`
+        :param y: Correct labels or target labels for `x`, depending if the attack is targeted
+               or not. This parameter is only used by some of the attacks.
+        :type y: `np.ndarray`
+        :return: An array holding the adversarial examples.
+        :rtype: `np.ndarray`
+        """
+        adv_x = copy.copy(x)
+
+        def minfun(x, args):  # minimize L2 norm
+            return np.sum(np.sqrt((x-args['orig'])**2))
+
+        def constraint_conf(x, args):  # constraint for confidence
+            pred = args['classifier'].predict(x.reshape(1, -1))[0, 0]
+            if args['class_zero']:
+                pred = 1.0-pred
+            return (pred-0.95).reshape(-1)
+
+        def constraint_unc(x, args):  # constraint for uncertainty
+            return (args['max_uncertainty']-(args['classifier'].predict_uncertainty(x.reshape(1, -1))).reshape(-1))[0]
+        bounds = []
+        # adding bounds, to not go away from original data
+        for i in range(np.shape(x)[1]):
+            bounds.append((self.minVal, self.maxVal))
+        for i in range(np.shape(x)[0]):  # go though data amd craft
+            # get properties for attack
+            max_uncertainty = self.unc_increase * \
+                self.classifier.predict_uncertainty(adv_x[i].reshape(1, -1))
+            class_zero = not self.classifier.predict(
+                adv_x[i].reshape(1, -1))[0, 0] < 0.5
+            init_args = {'classifier': self.classifier,
+                         'class_zero': class_zero, 'max_uncertainty': max_uncertainty}
+            cC = {'type': 'ineq', 'fun': constraint_conf, 'args': (init_args,)}
```

Comment:
Would it be possible to find a snake_case naming style variable name for `cC` and `cU`?